Templated secrets and configs

aaronlehmann opened the pull request:
This adds support for template expansion in secrets and configs, as added up stream in https://github.com/docker/swarmkit/pull/2133.

`SecretSpec` and `ConfigSpec` have a new optional `Templating` field. Currently, the only supported value is `golang` for go-style templates, but I imagine supporting other templating engines in the future, as Go-style templating is not especially familiar to Docker's target audience.

When templating is enabled, it's possible to expand information about the task and service (examples: `{{.Task.ID}}`, `{{env "VAR"}}`), and other secrets and configs available to the service (example: `{{secret "sometarget"}}`).

A few areas where this can be useful:

- Fill in information in a configuration file based on the task, service, or environment.
- Insert a secret such as a password into a configuration file, so the configuration file template can be managed separatly from the actual credential.
- Compose a secret from multiple secrets.

cc @diogomonica

Commit:
executor: Use a TemplatedDependencyGetter to support template expansion

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
api: Add Templating parameter to SecretSpec and ConfigSpec

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
integration-cli: Add secret/config templating tests

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

cpuguy83 commented:
Just so I understand, this is making it so you can template the actual secret/config data and effectively creates a dependencies between secrets/configs (assuming that the template is referencing a secret/config).

aaronlehmann commented:
> Just so I understand, this is making it so you can template the actual secret/config data and effectively creates a dependencies between secrets/configs (assuming that the template is referencing a secret/config).

I suppose you can say that, but you only secrets and configs already attached to the service may be referenced. So it effectively creates dependencies between services/configs, but cannot add new secret/config dependencies to a service.

cpuguy83 commented on integration-cli/docker_api_swarm_config_test.go:
```diff
@@ -116,3 +116,100 @@ func (s *DockerSwarmSuite) TestAPISwarmConfigsUpdate(c *check.C) {
+			Name: "golang",
+		},
+		Data: []byte("SERVICE_NAME={{.Service.Name}}\n" +
+			"{{secret \"referencedsecrettarget\"}}\n" +
```
If configs can pull in secrets, doesn't this have security consequences since configs are not on tmpfs?

justincormack replied:
that seems really complicated, why don't we move all configs to tmpfs? They should be accounted for under the container's memory usage so I don't think thats an issue.

aaronlehmann replied:
That would be fine with me. I don't have time to update the PR right now, but it's okay with me if someone wants to carry this.

aaronlehmann commented on integration-cli/docker_api_swarm_config_test.go:
```diff
@@ -116,3 +116,100 @@ func (s *DockerSwarmSuite) TestAPISwarmConfigsUpdate(c *check.C) {
+			Name: "golang",
+		},
+		Data: []byte("SERVICE_NAME={{.Service.Name}}\n" +
+			"{{secret \"referencedsecrettarget\"}}\n" +
```
I brought that up here, but there didn't seem to be concerns about it: https://github.com/docker/swarmkit/pull/2133#issuecomment-299989748

I'd be fine with restricting the ability to insert secrets into configs if people think that's the right thing to do.

cpuguy83 replied:
Yeah, I think it's a mistake to allow importing secrets into a config.
/cc @diogomonica

diogomonica replied:
Aren't configs also tmpFS?

aaronlehmann replied:
Nope. See https://github.com/docker/swarmkit/pull/2133#issuecomment-299989748

diogomonica replied:
Crap. And I LGTMed that? :(

diogomonica replied:
I think @cpuguy83 is right, unfortunately. This does break the fact that a `docker commit` will never keep secrets.

cyli replied:
I vaguely remember the outcome of that discussion being that if configs are flagged having a dependency on a secret, it was going to be treated as a secret?  It was left as an implementation detail for the future, though.

aaronlehmann replied:
I started working on flagging configs that depend on a secret so they can be stored in tmpfs. https://github.com/docker/swarmkit/pull/2286 is the swarmkit part of this. I will update this PR to make use of this.

aaronlehmann replied:
Updated the PR with new commits to store configs in the tmpfs secrets directory when they have secrets inserted inside them. Marking WIP because this depends on docker/swarmkit#2286.

aaronlehmann commented:
Rebased

Commit:
Store configs that contain secrets on tmpfs

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
docker-py test failures are unrelated (#33863)

aaronlehmann commented:
Swarmkit PR was merged. Rebased this one.

ping @diogomonica @cyli @cpuguy83

cyli commented:
Not sure if the lack of network-create event failures are related to the vendoring... but the config changes LGTM.

aaronlehmann commented:
I broke something in https://github.com/docker/swarmkit/pull/2310. I messed up the lifecycle on the watch server - it's supposed to be available on non-leader nodes as well. I'll fix it.

aaronlehmann commented:
Updated with fixed vendoring

cpuguy83 commented:
Design OK.

cpuguy83 commented on daemon/container_operations_unix.go:
```diff
@@ -138,41 +139,23 @@ func (daemon *Daemon) setupIpcDirs(c *container.Container) error {
-		return errors.Wrap(err, "unable to setup secret mount")
+	if !*hasSecretDir {
+		daemon.createSecretDir(c)
+		*hasSecretDir = true
```
Should this still unmount on error?

cpuguy83 commented on daemon/container_operations_unix.go:
```diff
@@ -257,19 +278,32 @@ func (daemon *Daemon) setupConfigDir(c *container.Container) (setupErr error) {
+			configRef.Sensitive = true
+			fPath = c.SensitiveConfigFilePath(*configRef.ConfigReference)
+			if !*hasSecretDir {
+				daemon.createSecretDir(c)
```
Check error

cpuguy83 commented on daemon/oci_linux.go:
```diff
@@ -732,14 +732,27 @@ func (daemon *Daemon) createSpec(c *container.Container) (*specs.Spec, error) {
+		}
+	}()
+
+	if err := daemon.setupSecretDir(c, &hasSecretDir); err != nil {
```
Can these functions return a bool instead of passing a bool pointer?

aaronlehmann replied:
I think it would make the code a bit messier, since the function would have to take a boolean argument and return a new value. The calling code would need logic to act on the returned value, and forgetting to do this would lead to errors. But I'm fine with making the change if you prefer it.

The other thing we could consider is moving `hasSecretDir` to `Daemon`, which might be cleaner overall.

Commit:
daemon: Check return value of createSecretDir

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

cpuguy83 commented:
I went ahead and rebased this. Still need to deal with configs on tmpfs. I can do this since I'd already started doing it for another issue.

cpuguy83 commented:
p.s. If we can review #35830 it will make it simpler to deal with the tmpfs change without rebase hell.

cpuguy83 committed:
Always mount configs with tmpfs

This makes configs and secrets behavior identical.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I've updated this to use mounts for configs.

aaronlehmann commented on daemon/container_operations_unix.go:
```diff
@@ -242,6 +243,25 @@ func (daemon *Daemon) createSecretDir(c *container.Container) error {
 
+func (daemon *Daemon) createConfigDir(c *container.Container) error {
+	localMountPath := c.ConfigsDirPath()
+	logrus.Debugf("secrets: setting up secret dir: %s", localMountPath)
```
I believe everything below here can be factored out into a common function.

aaronlehmann commented on daemon/container_operations_unix.go:
```diff
@@ -268,89 +304,78 @@ func (daemon *Daemon) cleanupSecretDir(c *container.Container) {
+	detachMounted(localMountPath)
+
+	if err := os.RemoveAll(localMountPath); err != nil {
+		logrus.Errorf("error cleaning up secret mount: %s", err)
```
`error cleaning up config mount`

aaronlehmann commented on daemon/container_operations_unix.go:
```diff
@@ -268,89 +304,78 @@ func (daemon *Daemon) cleanupSecretDir(c *container.Container) {
+		fPath := c.ConfigFilePath(*ref.ConfigReference)
 		if err := idtools.MkdirAllAndChown(filepath.Dir(fPath), 0700, rootIDs); err != nil {
-			return errors.Wrap(err, "error creating config path")
+			return errors.Wrap(err, "error creating config mount path")
```
Maybe "mount point" is the best terminology.

aaronlehmann commented on daemon/oci_linux.go:
```diff
@@ -816,27 +816,21 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
 	}
 
-	if err := daemon.setupConfigDir(c, &hasSecretDir); err != nil {
+	if err := daemon.setupConfigDir(c); err != nil {
```
Could we further simplify this by having only a single tmpfs mount for both secrets and configs? I don't think keeping them separate is buying us much.

cpuguy83 replied:
Possibly, but they will still need to be namespaced due to potential name collisions, and then the mounting behavior changed.
For now it's probably easier to keep them separate and we could look at a refactor later.

cpuguy83 replied:
Looking a little closer, maybe it's not so big a change.

cpuguy83 replied:
Windows makes this a little more tricky.

aaronlehmann reviewed (commented):
Thanks for doing this! I'm very happy to see this PR revived, and I think bringing the config/secret behavior into alignment makes sense.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -3333,6 +3333,18 @@ definitions:
         $ref: "#/definitions/Driver"
+      Templating:
+        description: "Templating driver, if applicable"
+        type: "object"
```
Like `Driver` above, this is a `Driver` type (also perhaps take some of the description from the `ConfigSpec` struct), something like;

```yaml
Templating:
  description: |
    Templating driver, if applicable.

    Templating controls whether and how to evaluate the config payload as a
    template. If no driver is set, no templating is used.
  $ref: "#/definitions/Driver"
```

thaJeztah commented on api/swagger.yaml:
```diff
@@ -3369,6 +3381,18 @@ definitions:
           config data.
         type: "string"
+      Templating:
+        description: "Templating driver, if applicable"
```
Same here:

```yaml
Templating:
  description: |
    Templating driver, if applicable.

    Templating controls whether and how to evaluate the config payload as a
    template. If no driver is set, no templating is used.
  $ref: "#/definitions/Driver"
```

thaJeztah commented on daemon/container_operations_unix.go:
```diff
@@ -235,8 +221,55 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
+
+func (daemon *Daemon) createConfigDir(c *container.Container) error {
+	localMountPath := c.ConfigsDirPath()
+	logrus.Debugf("secrets: setting up secret dir: %s", localMountPath)
```
s/secrets/configs/

thaJeztah commented on daemon/container_operations_unix.go:
```diff
@@ -235,8 +221,55 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
+	rootIDs := daemon.idMappings.RootPair()
+	// create tmpfs
+	if err := idtools.MkdirAllAndChown(localMountPath, 0700, rootIDs); err != nil {
+		return errors.Wrap(err, "error creating secret local mount path")
```
s/secret/config/

thaJeztah commented on daemon/container_operations_unix.go:
```diff
@@ -235,8 +221,55 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
+
+	tmpfsOwnership := fmt.Sprintf("uid=%d,gid=%d", rootIDs.UID, rootIDs.GID)
+	if err := mount.Mount("tmpfs", localMountPath, "tmpfs", "nodev,nosuid,noexec,"+tmpfsOwnership); err != nil {
+		return errors.Wrap(err, "unable to setup secret mount")
```
s/secret/config/

thaJeztah commented on container/container.go:
```diff
@@ -68,6 +68,13 @@ type ExitStatus struct {
+type ConfigReference struct {
+	*swarmtypes.ConfigReference
+	// Sensitive is set if this config should not be written to disk.
+	Sensitive bool
```
Silly question: should the templating driver also be aware of the content being sensitive or not? (just wondering if drivers may have to treat sensitive content different than non-sensitive)

thaJeztah commented on daemon/container_operations_unix.go:
```diff
@@ -245,74 +278,104 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
 			return errors.Wrap(err, "error setting ownership for config")
 		}
-
-		label.Relabel(fPath, c.MountLabel, false)
```
Should this be in this PR, or separate?

thaJeztah replied:
Ah, thanks; it just caught my eye 👍

thaJeztah commented on daemon/container_operations_unix.go:
```diff
@@ -245,74 +278,104 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
+	tmpfsOwnership := fmt.Sprintf("uid=%d,gid=%d", rootIDs.UID, rootIDs.GID)
+
+	// remount configs ro
+	if err := mount.Mount("tmpfs", localMountPath, "tmpfs", "remount,ro,"+tmpfsOwnership); err != nil {
```
Just some thoughts while reading this;

- should the `tmpfs` have a _size_ specified? (based on the size of the configs/secrets)?
- (off topic); would it be possible in future to have non-readonly configs (probably tricky)?  - thinking of applications that use a configuration, but may add runtime configuration to the initial configuration.

thaJeztah reviewed (changes requested):
left some nits, notes, and thoughts

cpuguy83 committed:
Merge configs/secrets in unix implementation

On unix, merge secrets/configs handling. This is important because
configs can contain secrets (via templating) and potentially a config
could just simply have secret information "by accident" from the user.
This just make sure that configs are as secure as secrets and de-dups a
lot of code.
Generally this makes everything simpler and configs more secure.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

thaJeztah commented:
Looks like CI is failing (note: different failures in some).

Also failure in docker-py;

```
16:29:23 _________________ ServiceTest.test_create_service_with_config __________________
16:29:23 /docker-py/tests/integration/api_service_test.py:543: in test_create_service_with_config
16:29:23     assert container is not None
16:29:23 E   AssertionError: assert None is not None
16:29:23 _____________ ServiceTest.test_create_service_with_unicode_config ______________
16:29:23 /docker-py/tests/integration/api_service_test.py:568: in test_create_service_with_unicode_config
16:29:23     assert container is not None
16:29:23 E   AssertionError: assert None is not None
16:29:23 ========= 2 failed, 277 passed, 5 skipped, 4 xfailed in 335.08 seconds =========
```

cpuguy83 commented:
Yes, will have to take a look. Saw something similar in another PR where I
moved around configs.

On Thu, Jan 18, 2018 at 9:47 AM, Sebastiaan van Stijn <
notifications@github.com> wrote:

> Looks like CI is failing (note: different failures in some).
>
> Also failure in docker-py;
>
> 16:29:23 _________________ ServiceTest.test_create_service_with_config __________________
> 16:29:23 /docker-py/tests/integration/api_service_test.py:543: in test_create_service_with_config
> 16:29:23     assert container is not None
> 16:29:23 E   AssertionError: assert None is not None
> 16:29:23 _____________ ServiceTest.test_create_service_with_unicode_config ______________
> 16:29:23 /docker-py/tests/integration/api_service_test.py:568: in test_create_service_with_unicode_config
> 16:29:23     assert container is not None
> 16:29:23 E   AssertionError: assert None is not None
> 16:29:23 ========= 2 failed, 277 passed, 5 skipped, 4 xfailed in 335.08 seconds =========
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/33702#issuecomment-358668097>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAwxZrq2PkcV94e8Be17YJtexxQFOUE0ks5tL1lvgaJpZM4N7oq7>
> .
>



-- 


- Brian Goff

cpuguy83 commented:
This should be ready.

thaJeztah commented:
Some things left;

- the `Templating` option should be gated by API version
- the new option should be in the API version history
- a follow-up PR for the CLI

TBD:

- do we want the default templating driver to be named "golang", or should we name it "default"? ("golang" feels like an implementation detail); I _think_ I'd prefer "default", but open to input.

Improvements;

Using a non-existing secret/config makes deploying the task fail (note: only on secrets/configs; non-existing environment variables are silently ignored, which is probably ok? or do we want to error on those as well?)

The error-message in that situation is a bit cryptic, and could be improved a bit (fine for a follow-up):

```
starting container failed: unable to get config from config store: failed to expand templated config 6l98oxwkxgaegswl411kbgnr7: template: expansion:1:25: executing "expansion" at <secret "nosuchsecret">: error calling secret: secret target nosuchsecret not found
```

cpuguy83 commented:
> the Templating option should be gated by API version

We don't generally do this, do we?

> do we want the default templating driver to be named "golang"
Yes, I think so because this is somewhat specific to go and trying to feed the same template into some other templating engine would likely fail.

thaJeztah commented:
> We don't generally do this, do we?

I know at least we do it for the `AutoRemove` option; if we didn't gate it, effectively we're changing older API versions (because the option isn't ignored). This is how we declare our "open schema" API;

> The API uses an open schema model, which means server may add extra properties to responses. *Likewise, the server will ignore any extra query parameters and request body properties*. When you write clients, you need to ignore additional properties in responses to ensure they do not break when talking to newer daemons.

vdemeester commented on container/container_windows.go:
```diff
@@ -101,24 +102,20 @@ func (container *Container) CreateConfigSymlinks() error {
+// TODO: Right now Windows doesn't really have a "secure" storage for secrets,
+// however some configs may contain secrets. Once secure storage is worked out,
+// configs and secret handling should be merged.
+func (container *Container) ConfigMounts() []Mount {
```
Won't this fail to validate (golint 😛)

vdemeester reviewed (commented):
With adding a way to ignore the templating if the API version is lower than the current one (don't remember the exact number), SGTM 👍

cpuguy83 committed:
Error out on secret/config templates for older API

Makes sure if the user specifies an older API version that we don't pass
through templating options for versions that templating was not
supported.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

vdemeester reviewed (approved):
LGTM :lion:

thaJeztah commented on api/server/router/swarm/cluster_routes.go:
```diff
@@ -372,6 +372,10 @@ func (sr *swarmRouter) createSecret(ctx context.Context, w http.ResponseWriter,
 	}
+	version := httputils.VersionFromContext(ctx)
+	if secret.Templating != nil && versions.LessThan(version, "1.36") {
+		return errdefs.InvalidParameter(errors.Errorf("secret templating is not supported on the specified API version: %s", version))
```
Hm initially thought we'd _ignore_ the property (as in: "old API doesn't know about this option, so just ignores it"), but I guess it's ok

thaJeztah commented on api/server/router/swarm/cluster_routes.go:
```diff
@@ -372,6 +372,10 @@ func (sr *swarmRouter) createSecret(ctx context.Context, w http.ResponseWriter,
 		return err
 	}
+	version := httputils.VersionFromContext(ctx)
+	if secret.Templating != nil && versions.LessThan(version, "1.36") {
```
oh, dang! this needs to be 1.37 now (https://github.com/moby/moby/pull/33922 was merged)

thaJeztah replied:
I'm ok with fixing that in a follow-up

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
Ignoring codecode/patch;

> Missing base report.
> Unable to compare commits because the base of the compare did not upload a coverage report.
